migration-strategies/piercing/app/fragments/login/src/root.tsx

```diff
@@ -3,123 +3,95 @@ import {
   component$,
   useRef,
   useStore,
-  useStyles$,
   useStylesScoped$,
 } from "@builder.io/qwik";
 import { dispatchPiercingEvent } from "piercing-library";
 
 import styles from "./root.css?inline";
 
-export const Root = component$(() => {
-  const possibleInputErrors = {
-    noSpaces: "noSpaces",
-    noSpecialCharacters: "noSpecialCharacters",
-    maxLength: "maxLength",
-    emptyInput: "emptyInput",
-  } as const;
+const enum InputErrorMessages {
+  noSpaces = "Spaces are not allowed",
+  noSpecialCharacters = "Special characters are not allowed",
+  maxLength = "The provided username is too long, it needs not to be longer than 20 characters",
+  emptyInput = "Please provide a username",
+}
+interface LoginState {
+  animatePassword: boolean;
+  usernameError: InputErrorMessages | null;
+  usernameTouched: boolean;
+  username: string;
+  password: string;
+  loading: boolean;
+}
 
-  const inputErrorMessages: Record<
-    typeof possibleInputErrors[keyof typeof possibleInputErrors],
-    string
-  > = {
-    [possibleInputErrors.noSpaces]: "Spaces are not allowed",
-    [possibleInputErrors.noSpecialCharacters]:
-      "Special characters are not allowed",
-    [possibleInputErrors.maxLength]:
-      "The provided username is too long, it needs not to be longer than 20 characters",
-    [possibleInputErrors.emptyInput]: "Please provide a username",
-  };
+export const getUsernameInputError = (
+  input: string,
+  inputTouched: boolean
+): InputErrorMessages | null => {
+  if (input.includes(" ")) return InputErrorMessages.noSpaces;
+  if (/[^a-zA-Z0-9]/.test(input)) return InputErrorMessages.noSpecialCharacters;
+  if (input.length > 20) return InputErrorMessages.maxLength;
+  if (inputTouched && !input) return InputErrorMessages.emptyInput;
+  return null;
+};
 
+export const Root = component$(() => {
   const ref = useRef();
   useStylesScoped$(styles);
 
-  const inputsDetails = useStore<{
-    animatePasswordInput: boolean;
-    usernameInputError:
-      | null
-      | typeof possibleInputErrors[keyof typeof possibleInputErrors];
-    usernameInputTouched: boolean;
-  }>({
-    animatePasswordInput: false,
-    usernameInputError: null,
-    usernameInputTouched: false,
-  });
-
-  const userData = useStore({
+  const state = useStore<LoginState>({
```

Comment:
In simple components like this I don't think there is much value in having multiple "stores".
If we wanted the code to be reusable, then I think we would move it into a custom hook instead.

Reply:
yeah.... I was just trying to have a logical distinction of the different states for the component (I wasn't doing it for performance or reusability reasons), it was sort of an attempt to make things more separated/clear, but I guess it may just raise more confusion than anything else 😖 

I'm totally ok with having only single "stores" 🙂